tiledb/sm/serialization/array_schema.cc

```diff
@@ -304,110 +305,166 @@ Status dimension_to_capnp(
+      break;
+    }
+    case Datatype::INT16: {
+      auto val = tile_extent_reader.getInt16();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT16: {
+      auto val = tile_extent_reader.getUint16();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::INT32: {
+      auto val = tile_extent_reader.getInt32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT32: {
+      auto val = tile_extent_reader.getUint32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::DATETIME_YEAR:
+    case Datatype::DATETIME_MONTH:
+    case Datatype::DATETIME_WEEK:
+    case Datatype::DATETIME_DAY:
+    case Datatype::DATETIME_HR:
+    case Datatype::DATETIME_MIN:
+    case Datatype::DATETIME_SEC:
+    case Datatype::DATETIME_MS:
+    case Datatype::DATETIME_US:
+    case Datatype::DATETIME_NS:
+    case Datatype::DATETIME_PS:
+    case Datatype::DATETIME_FS:
+    case Datatype::DATETIME_AS:
+    case Datatype::TIME_HR:
+    case Datatype::TIME_MIN:
+    case Datatype::TIME_SEC:
+    case Datatype::TIME_MS:
+    case Datatype::TIME_US:
+    case Datatype::TIME_NS:
+    case Datatype::TIME_PS:
+    case Datatype::TIME_FS:
+    case Datatype::TIME_AS:
+    case Datatype::INT64: {
+      auto val = tile_extent_reader.getInt64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT64: {
+      auto val = tile_extent_reader.getUint64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT32: {
+      auto val = tile_extent_reader.getFloat32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT64: {
+      auto val = tile_extent_reader.getFloat64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    default:
+      throw std::logic_error(
+          "From tile_extent_from_capnp: Default branch of switch statement "
+          "should not be reached.");
+  }
+  return tile_extent;
+}
+
+shared_ptr<Dimension> dimension_from_capnp(
+    const capnp::Dimension::Reader& dimension_reader) {
+  // datatype
+  Datatype dim_type;
+  Status s_datatype_enum =
+      datatype_enum(dimension_reader.getType().cStr(), &dim_type);
+  if (!s_datatype_enum.ok()) {
+    throw std::runtime_error(
+        "From dimension_from_capnp: Could not process datatype obtained from "
+        "the capnp dimension reader object.");
+  }
+
+  // When the datatype_size function returns 0, we know that the datatype is
+  // invalid.
+  auto coord_size = datatype_size(dim_type);
+  if (coord_size == 0) {
+    throw std::runtime_error(
+        "From dimension_from_capnp: Input datatype " +
+        std::to_string(static_cast<uint8_t>(dim_type)) + " is invalid.");
```

Comment:
See above comments.

Ask yourself: is `dim_type` considered to have been validated or not?